src/util/sharing_node.h

```diff
@@ -12,340 +12,499 @@ Author: Daniel Poetzl
 #ifndef CPROVER_UTIL_SHARING_NODE_H
 #define CPROVER_UTIL_SHARING_NODE_H
 
-#include <list>
+#define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <forward_list>
 #include <map>
 #include <memory>
-#include <cassert>
+#include <type_traits>
 
 #include "invariant.h"
+#include "make_unique.h"
 
-#define _sn_assert(b) INVARIANT(b, "Sharing-node internal invariant")
+#define _sn_assert(b) INVARIANT(b, "Sharing node internal invariant")
 //#define _sn_assert(b)
 
+#define _sn_type_par_decl                                                      \
+  template <typename keyT,                                                     \
+            typename valueT,                                                   \
+            typename equalT = std::equal_to<keyT>>
+
+#define _sn_type_par_def                                                       \
+  template <typename keyT, typename valueT, typename equalT>
+
+#define _sn_type_args keyT, valueT, equalT
+
 template <class T>
 const T *as_const(T *t)
 {
   return t;
 }
 
-template <
-  class keyT,
-  class valueT,
-  class predT=std::equal_to<keyT>,
-  bool no_sharing=false>
-class sharing_nodet
+template <typename baseT, typename dataT>
+struct delete_through_base_ptrt
+{
+  static_assert(std::is_base_of<baseT, dataT>::value, "");
+
+  void operator()(baseT *bp)
+  {
+    _sn_assert((!std::is_same<baseT, dataT>::value));
+
+    dataT *dp = static_cast<dataT *>(bp);
+    delete dp;
+  }
+};
+
+template <typename baseT, typename dataT = baseT>
+void detach_shared_data(
+  std::shared_ptr<baseT> &data,
+  const std::shared_ptr<baseT> &empty_data)
+{
+  static_assert(std::is_base_of<baseT, dataT>::value, "");
+
+  _sn_assert(data.use_count() > 0);
+
+  if(std::is_same<baseT, dataT>::value)
+  {
+    if(data == empty_data)
+    {
+      data = std::make_shared<dataT>();
+    }
+    else if(data.use_count() > 1)
+    {
+      data = std::make_shared<dataT>(*data);
+    }
+  }
+  else
+  {
+    typedef delete_through_base_ptrt<baseT, dataT> deletet;
+
+    if(data == empty_data)
+    {
+      data.reset(new dataT(), deletet());
+    }
+    else if(data.use_count() > 1)
+    {
+      auto p = std::static_pointer_cast<dataT>(data);
```

Comment:
You assume that `data` is pointing to an instance of `dataT` here. Can you add an assertion for that?

Reply:
Also, please leave a comment explaining why you're using static rather than dynamic cast for this.